File: apple/Sources/ReactNativeWatchOS/Modifiers/ToolbarTitleDisplayMode.swift
import SwiftUI

/// SwiftUI `.toolbarTitleDisplayMode(_:)` (watchOS 10+). Controls how the
/// navigation title is displayed: `automatic`, `inline`, or `large`. Returns
/// the view unchanged on watchOS 9.
enum RNWToolbarTitleDisplayModeModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("toolbarTitleDisplayMode") { view, params, _ in
            if #available(watchOS 10.0, *) {
                let mode: ToolbarTitleDisplayMode
                switch params.string("mode") {
                case "inline": mode = .inline
                case "large":  mode = .large
                default:       mode = .automatic
                }
                return AnyView(view.toolbarTitleDisplayMode(mode))
            }
            return view
        }
    }
}

File: src/modifiers/toolbarTitleDisplayMode.ts
import { createModifier } from './createModifier';

/// SwiftUI `ToolbarTitleDisplayMode` values.
export type ToolbarTitleDisplayModeValue = 'automatic' | 'inline' | 'large';

export interface ToolbarTitleDisplayModeParams {
  /// `'automatic'` | `'inline'` | `'large'`.
  mode: ToolbarTitleDisplayModeValue;
}

/// SwiftUI `.toolbarTitleDisplayMode(_:)` (watchOS 10+). Controls how the
/// navigation title is displayed. No-op on watchOS 9.
export function toolbarTitleDisplayMode(
  mode: ToolbarTitleDisplayModeValue
): ReturnType<typeof createModifier>;
export function toolbarTitleDisplayMode(
  params: ToolbarTitleDisplayModeParams
): ReturnType<typeof createModifier>;
export function toolbarTitleDisplayMode(
  a: ToolbarTitleDisplayModeParams | ToolbarTitleDisplayModeValue
) {
  if (typeof a === 'string') {
    return createModifier('toolbarTitleDisplayMode', { mode: a });
  }
  return createModifier('toolbarTitleDisplayMode', a);
}
